### PR TITLE
Begin support for COINHSL

### PR DESCRIPTION
**Pull request by bjack205**
This allows for users to use the COINHSL package, which provides a suite of solvers for real data types, including ma97 and ma57. This PR adds some machinery to handle sources from both the normal library and the coinhsl package. I can successfully run ma97 from both libraries. ma57, however, crashed when I tried running the coinhsl version so I've disable it for now.

I'm hoping to add in the other solvers that ship with with package.

I'll add instructions in the repo for support for this once we add at least one more solver and have tested it on other machines.

*NOTE*: this adds onto the change to the build structure in #66

### CODE REVIEW

File: src/hsl_ma57.jl [hunk 1]
```diff
@@ -10,6 +10,10 @@ export Ma57Exception
 
 const Ma57Data = Union{Float32, Float64}
 
+const ma57alg = getalg(:hsl_ma57)
+const ma57types = getdatatypes(ma57alg)
+const libma57 = getlib(ma57alg)
```
Comment (dpo) [inline src/hsl_ma57.jl:15]: could we keep the name libhsl_ma57 here, so as to minimize changes? Same for ma97.

Comment (bjack205) [reply src/hsl_ma57.jl]: Fixed. If `libhsl_97` is defined it gets the value of `libcoinhsl`.

**Comment by bjack205** [comment]
This has also been updated with the most recent changes.

**Comment by amontoison** [comment]
@JSOBot runtests

File: deps/build_coinhsl.jl [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+archive = coinhsl_archive
+if ver.ext == ".tar.gz"
+    run(`tar -zxf $archive -C $builddir`)
+elseif ver.ext == ".zip"
+    run(`unzip -o $archive -d $builddir`)
+end
+
+name = getname(ver, false) 
+cd("$builddir/$name")
+download("http://glaros.dtc.umn.edu/gkhome/fetch/sw/metis/OLD/metis-4.0.3.tar.gz",joinpath(pwd(), "metis-4.0.3.tar.gz"))
```
Comment (dpo) [inline deps/build_coinhsl.jl:11]: We don't need to download METIS, it's already supplied by METIS4_jll.

File: deps/build.jl [hunk 3]
```diff
@@ -56,15 +56,27 @@ const hsl_ma57_patch = joinpath(hsl_ma57_path, "get_factors.patch")
...
 
+##############################
+# COINHSL
+##############################
+const hsl_coin_versions = [
+    # HSLVersion("coinhsl", "2021.05.05", "f77ad752a37de8695c02c81bcc503674af76689d40a9864b6f2a7a790c3efc95", ".tar.gz"),  # this has a problem with the install-sh script
+    HSLVersion("coinhsl", "2021.05.05", "62c7e18ff22b977afa442db97d791f31359ab4a4f5a027f315cace211a24fbe9", ".zip")
+]
+ENV["HSL_COIN_PATH"] = ENV["HSL_MA97_PATH"]
+const hsl_coin_path = haskey(ENV, "HSL_COIN_PATH") ? ENV["HSL_COIN_PATH"] : joinpath(@__DIR__, "downloads")
```
Comment (dpo) [inline deps/build.jl]: Not sure I understand this. The `HSL_COIN_PATH` was just defined at the line above.

File: deps/build.jl [hunk 4]
```diff
@@ -56,15 +56,27 @@ const hsl_ma57_patch = joinpath(hsl_ma57_path, "get_factors.patch")
...
+hsl_coin_version = findversion(hsl_coin_versions, hsl_coin_path)
```
Comment (dpo) [inline deps/build.jl]: When I downloaded CoinHSL, I obtained `coinhsl.tar.gz`, the version number wasn't part of the name (contrary to the MA57 and 97 archives). But `findversion()` will look for `coinhsl-2021.05.05.tar.gz`, won't it? (NB: the archive untars into the `coinhsl-2021.05.05` directory.)

File: deps/build_coinhsl.jl [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+elseif ver.ext == ".zip"
+    run(`unzip $archive -d $builddir`)
+end
+
+name = getname(ver, false) 
+cd("$builddir/$name")
+download("http://glaros.dtc.umn.edu/gkhome/fetch/sw/metis/OLD/metis-4.0.3.tar.gz",joinpath(pwd(), "metis-4.0.3.tar.gz"))
+run(`./configure --prefix=$usrdir F77=gfortran CFLAGS=-fPIC FFLAGS="-fPIC -fopenmp" FCFLAGS="-fPIC -fopenmp" --with-blas="-L$libopenblas_dir -lopenblas"`)
+run(`make install`)
+run(`gfortran -fPIC -shared -Wl,$all_load $libdir/lib$(ver.algname).a -L$libopenblas_dir -lopenblas -lgomp -Wl,$noall_load -o $libdir/lib$(ver.algname).$so`)
```
Comment (dpo) [inline deps/build_coinhsl.jl:14]: Now the compiler is supplied in an environment variable.

File: src/HSL.jl [hunk 6]
```diff
@@ -15,20 +15,87 @@ function __init__()
   end
 end
 
+const HASCOINHSL = @isdefined libcoinhsl
+
+struct HSLAlgorithm
+  name::Symbol
+  datatypes::Vector{DataType}
+  coinhsltypes::Vector{DataType}  # types provided by libcoinhsl, if any
```
Comment (dpo) [inline src/HSL.jl:26]: Why is this packaged together with, e.g., standalone MA57 or MA97?

**Review by dpo — COMMENTED** [review]
I have to say I'm a bit lukewarm about this. CoinHSL only provides s and d versions of the subroutines, which introduces quite a bit of complication.

Also can I please ask you to properly rebase this PR against the latest `main` (NB: rebase instead of merge)? I tried to do it myself but I want to be sure that I don't accidentally omit something you added in here.

**Comment by bjack205** [comment]
I'm fine if you'd prefer to not merge this PR. It does add quite a bit of complexity having to manage which library to pull the algorithms from, and as you said, COINHSL only provides support for real data types.

**Pull request closed without merging**